### PR TITLE
Fix OOB issues in Nef_*

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes
Add bound checks in IO functions of Ned_2/S2/3, and don't crash the program if there is an error but simply return witth a cerr.

## Release Management

* Affected package(s):Nef_2, Nef_S2, Nef_3

### CODE REVIEW

File: Nef_2/include/CGAL/Nef_2/PM_io_parser.h [hunk 1]
```diff
@@ -313,13 +340,26 @@ template <typename PMDEC>
 void PM_io_parser<PMDEC>::read()
 {
   if ( !check_sep("Plane_map_2") )
-    CGAL_error_msg("PM_io_parser::read: no embedded_PM header.");
+  {
+    std::cerr<<"PM_io_parser::read: no embedded_PM header."<<std::endl;
```
Comment (sloriot) [inline Nef_2/include/CGAL/Nef_2/PM_io_parser.h:344]: why removing the CGAL macros?

Comment (maxGimeno) [reply Nef_2/include/CGAL/Nef_2/PM_io_parser.h:344]: They call std::abort and crash the program.

Comment (sloriot) [reply Nef_2/include/CGAL/Nef_2/PM_io_parser.h:344]: Except if there is a verbose mode, nothing should be written to std::cerr. If there is no verbose mode maybe CGAL_warning_msg can be used.

File: Nef_3/include/CGAL/Nef_3/SNC_io_parser.h [hunk 2]
```diff
@@ -1400,24 +1400,47 @@ template <typename EW>
 void SNC_io_parser<EW>::read()
 {
   if ( !check_sep("Selective Nef Complex") )
-    CGAL_error_msg("SNC_io_parser::read: no SNC header.");
+  {
+    CGAL_warning_msg(false, "SNC_io_parser::read: no SNC header.");
+    returnl
```
Comment (lrineau) [inline Nef_3/include/CGAL/Nef_3/SNC_io_parser.h:1405]: Cannot compile.
```suggestion
    return;
```

File: Nef_2/include/CGAL/Nef_2/PM_io_parser.h [hunk 3]
```diff
@@ -267,14 +276,32 @@ bool PM_io_parser<PMDEC>::read_face(Face_handle f)
...
   if ( !(in >> n) || !check_sep("{") ) return false;
   if ( !(in >> ei) || !check_sep(",") ) return false;
-  if (ei >= 0) f->set_halfedge(Halfedge_of[ei]);
+  if (ei >= 0 && ei < en)
+  {
+    f->set_halfedge(Halfedge_of[ei]);
+  }
+  else
+  {
+    in.clear(std::ios_base::badbit);
```
Comment (lrineau) [inline Nef_2/include/CGAL/Nef_2/PM_io_parser.h:285]: I would prefer:
```suggestion
    in.setstate(std::ios_base::badbit);
```
I find strange to call `clear` to set a bit, and `clear` also has the defect that it clears all the other bits.

File: Nef_3/include/CGAL/Nef_3/SNC_io_parser.h (hunk 2 above)
Comment (GilesBathgate) [reply Nef_3/include/CGAL/Nef_3/SNC_io_parser.h:1405]: Doh!

**Comment by maxGimeno** [comment]
Errors:
 - https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.3-Ic-46/Nef_2/TestReport_lrineau_ArchLinux.gz

File: Nef_3/include/CGAL/Nef_3/SNC_io_parser.h [hunk 4]
```diff
@@ -1579,7 +1579,7 @@ read_vertex(Vertex_handle vh) {
   vh->sncp() = this->sncp();
 
   in >> index;
-  if(index >= int(en))
+  if(index >= int((int)(en)))
```
Comment (GilesBathgate) [inline Nef_3/include/CGAL/Nef_3/SNC_io_parser.h:1582]: isn't this a double cast? could be written as:
`(int)en`, instead of `int((int)en)`?

**Comment by maxGimeno** [comment]
Success in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-55.shtml

**Comment by jubalh** [comment]
https://github.com/CGAL/cgal/issues/5345 states that this PR fixes security issues. Would you mind commenting the CVE numbers which this PR is fixing?

**Comment by GilesBathgate** [comment]
@jubalh I think these would be instances of [CWE-606](https://cwe.mitre.org/data/definitions/606.html)

**Comment by lrineau** [comment]
See https://talosintelligence.com/vulnerability_reports/TALOS-2020-1225 for the details.

**Comment by jubalh** [comment]
Thanks. I'll comment the CVE numbers here so that it's easier for others to search for them.
CVE-2020-28601-CVE-2020-28636, CVE-2020-35628-CVE-2020-35636

**Pull request merged**